### PR TITLE
RFC: add GLUON_MINIFY flag to allow skipping the minification process

### PR DESCRIPTION
**Pull request by lemoer**
This is just an idea, which might be helpful for development images.

What do you think?

### CODE REVIEW

**Comment by christf** [comment]
I like the idea. Currently I manually copy over the source, then fix it on the node and copy it back. Not minifying would eliminate that first step.

Do we need a new flag for this or should the logic be triggered by GLUON_DEBUG?

**Comment by rotanid** [comment]
having it as a separate flag may allow to use it on smaller devices than with complete GLUON_DEBUG
but maybe it should be enabled automatically, if gluon_debug is active.

shouldn't it also "stop" the shell minification, not only lua?

**Comment by lemoer** [comment]
I would suggest, that we enable it by default on devies with large storages as x86-generic for example. This should not have any disadvantages, but would make developing easier.

@rotanid: The commit also adjusts the `gluon-status-page` package to be built with the non-uglified `status-page.js`. Is this, what you are referring to? Or is there other Code, which is minified?

**Comment by rotanid** [comment]
i think your suggestion is not good, because if you only test on x86, you won't catch problems with minification then.

i saw your status page change, but i remember there are also shell script minifications, maybe not in the gluon repo but in packages..

**Comment by lemoer** [comment]
> i saw your status page change, but i remember there are also shell script minifications, maybe not in the gluon repo but in packages..

I don't know about them, but even if so, it is more difficult to deactivate minification for packages which are outside of gluon and are not built on base of `gluon.mk`. This would have to be patched in the external package repo itself.

> i think your suggestion is not good, because if you only test on x86, you won't catch problems with minification then.

I see your point. Then I would suggest to introduce another image besides `x86-generic`, which is called `x86-dev` which has no minifications. What do you think?

**Comment by CodeFetch** [comment]
I think it should be automatically enabled on GLUON_DEBUG, but it should also be possible to skip the minification only, because debug images are too big. Can't we add a flag to OpenWrt to skip minification, too? I remember such flags for skipping the binary stripping.

**Comment by NeoRaider** [comment]
Generally the feature seems useful, but the implementation is broken: Environment variables must not directly affect the build, otherwise PKG_CONFIG_DEPENDS does not work; this must be added to openwrt/.config through `package/gluon-core/Config.in` instead.

Also, I agree with @CodeFetch that this seems like a good addition for GLUON_DEBUG rather than adding a new env variable.

**Comment by rotanid** [comment]
@NeoRaider it is NOT sufficient to extend GLUON_DEBUG, because you can't  put such a build on most devices because of flash size issues. what you suggest is NOT what codefetch wrote and what lemoer wants.

**Comment by NeoRaider** [comment]
I see. I'm not happy about adding too many different flags, but I guess it can't be avoided in this case.

**Comment by rotanid** [comment]
@lemoer can you implement the change neoraider suggested 3 comments above?

**Comment by lemoer** [comment]
Tbh, I am not really sure how to implement his suggestion. If I understood it correctly, the flag should be added like this:
``` diff
diff --git a/package/gluon-core/Config.in b/package/gluon-core/Config.in
index 3f381f92..6abbc2ce 100644
--- a/package/gluon-core/Config.in
+++ b/package/gluon-core/Config.in
@@ -347,3 +347,6 @@ config GLUON_SPECIALIZE_KERNEL
        select KERNEL_IP_NF_NAT
        select KERNEL_IP_NF_TARGET_MASQUERADE
        select KERNEL_IP_NF_TARGET_REDIRECT
+
+config GLUON_DO_NOT_MINIFY
+       bool
```

But as it is still desired to use this flag as an environment variable from `make`, how do I pass the information from the environment variable to this build variable? Where is the right place to do this? Or perhaps I am completely misunderstanding this?

**Comment by NeoRaider** [comment]
- The config symbol in `Config.in` needs a human-readable label (symbols without a label can only be enabled by the `select` of another symbol)
- All generic .config symbols are handled in `targets/generic`, where GLUON_DEBUG is handled as well

**Comment by lemoer** [comment]
I hope, it's correct now...

**Comment by NeoRaider** [comment]
- Please rebase, I just replaced `envtrue` with a more intuitive `istrue` helper (see latest commits in master)
- Check for the config symbol using:
```make
ifdef CONFIG_GLUON_DO_NOT_MINIFY
...
endif
```
(or ifndef instead of ifdef)
- `PKG_CONFIG_DEPENDS += CONFIG_GLUON_DO_NOT_MINIFY`
- Better symbol label: "Disable minification of Gluon scripts"
- It is uncommon to have a negated .config symbol, the following might be better:
```
config GLUON_MINIFY
  bool "Minify Gluon scripts"
  default y
```

**Comment by rotanid** [comment]
@lemoer apart from a rebase, do you intend to continue with this following neoraiders input?

**Comment by lemoer** [comment]
@rotanid done.

File: Makefile [hunk 1]
```diff
@@ -55,7 +55,8 @@ endef
 GLUON_VARS = \
 	GLUON_RELEASE GLUON_REGION GLUON_MULTIDOMAIN GLUON_DEBUG GLUON_DEPRECATED GLUON_DEVICES \
 	GLUON_TARGETSDIR GLUON_PATCHESDIR GLUON_TMPDIR GLUON_IMAGEDIR GLUON_PACKAGEDIR \
-	GLUON_SITEDIR GLUON_RELEASE GLUON_BRANCH GLUON_LANGS GLUON_BASE_FEEDS BOARD SUBTARGET
+	GLUON_SITEDIR GLUON_RELEASE GLUON_BRANCH GLUON_LANGS GLUON_BASE_FEEDS BOARD SUBTARGET \
+	GLUON_MINIFY
```
Comment (NeoRaider) [inline Makefile:59]: Add `GLUON_MINIFY ?= 1` with the other defaults, and move next to GLUON_DEBUG in the GLUON_VARS list.

File: docs/user/getting_started.rst [hunk 2]
```diff
@@ -212,6 +212,12 @@ GLUON_DEBUG
   Setting ``GLUON_DEBUG=1`` will provide firmware images including debugging symbols usable with GDB or
   similar tools. Requires a device or target with at least 16 MB of flash space, e.g. `x86-64`. Unset by default.
 
+GLUON_MINIFY
+  Setting ``GLUON_MINIFY`` will minify scripts during the build process. This flag is enabled by
```
Comment (NeoRaider) [inline docs/user/getting_started.rst:216]: Should note that valid values are 0 and 1.

File: docs/user/getting_started.rst [hunk 3]
```diff
@@ -212,6 +212,12 @@ GLUON_DEBUG
...
+  default. Disabling this flag is handy if human readable scripts on the devices are desired for
+  development purposes. Be aware, that this will increase the size of the resulting images and
```
Comment (NeoRaider) [inline docs/user/getting_started.rst:218]: Remove this comma

**Comment by lemoer** [comment]
@NeoRaider Done.

File: docs/user/getting_started.rst [hunk 4]
```diff
@@ -212,6 +212,12 @@ GLUON_DEBUG
   Setting ``GLUON_DEBUG=1`` will provide firmware images including debugging symbols usable with GDB or
   similar tools. Requires a device or target with at least 16 MB of flash space, e.g. `x86-64`. Unset by default.
 
+GLUON_MINIFY
+  Setting ``GLUON_MINIFY=1`` will minify scripts during the build process. This flag is enabled by
```
Comment (NeoRaider) [inline docs/user/getting_started.rst:216]: As this is enabled by default, the docs should be about `GLUON_MINIFY=0`.

**Comment by lemoer** [comment]
Updated.

**Pull request merged**